File: examples/clojure/07-fp-features/main.clj
;; FP Features in Clojure
;; ======================
;; Pure functions, immutable data, higher-order functions

(ns main)

;; ── Pure functions ─────────────────────────────

(println "--- Pure functions ---")

(def add (fn [a b] (+ a b)))
(def multiply (fn [a b] (* a b)))

;; Function as value
(println "add(5, 3) =" (add 5 3))

;; Higher-order function
(defn apply-twice [f x] (f (f x)))
(println "apply-twice(x+1, 5) =" (apply-twice inc 5))


;; ── Higher-order functions (map, filter, reduce) ───────

(println "\n--- Higher-order functions ---")

(def numbers [1 2 3 4 5 6 7 8 9 10])

;; map
(def squares (map #(* % %) numbers))
(println "Squares:" squares)

;; filter
(def evens (filter even? numbers))
(println "Evens:" evens)

;; reduce
(def total (reduce + 0 numbers))
(println "Sum:" total)

;; find
(def first-even (first (filter even? numbers)))
(println "First even:" first-even)

;; any/some
(def has-odd (some odd? numbers))
(println "Has odd? " has-odd)

(def all-positive? (every pos? numbers))
(println "All positive? " all-positive?)


;; ── Closures (fn, let, loop) ─────────────────────

(println "\n--- Closures ---")

(def make-multiplier (fn [factor]
                (fn [x] (* x factor))))

(def times-three (make-multiplier 3))
(def times-five (make-multiplier 5))

(println "times-three(7) =" (times-three 7))
(println "times-five(8) =" (times-five 8))


;; ── Partial application (partial) ─────────────────────

(println "\n--- Partial application (partial)")

(def add-one (partial + 1))
(def times-two (partial * 2))

(println "add-one(5) =" (add-one 5))
(println "times-two(5) =" (times-two 5))


;; ── Function composition (comp, juxt) ───────────────

(println "\n--- Function composition (comp, juxt)")

(def shout (comp str/upper str/trim))
(def greet (partial str "Hello, "))

(println "shout(\"  alice  \") =" (shout "  alice  "))
(println "greet(\"Bob\") =" (greet "Bob"))

;; juxt (apply multiple functions)
(def describe-number (juxt name count))
(println "describe-number(42) =" (describe-number 42))


;; ── Destructuring ─────────────────────────────────────

(println "\n--- Destructuring")

(defn process-point [[x y]]
  (+ x y))

(def point [3 4])
(println "Point sum:" (process-point point))

;; Destructuring in function params
(defn destructure-map [{:keys [name age]}]
  (str name " is " age " years))

(def person-map {:name "Alice", :age 30})
(println (destructure-map person-map))


;; ── Lazy sequences ───────────────────────────────

(println "\n--- Lazy sequences")

(def naturals (range))
(def take-ten (take 10 naturals))

(println "First 10 naturals:" (take-ten))


;; ── Transducer ───────────────────────────────────

(println "\n--- Transducer (composition for reduce)")

(def xf-sum (map #(* % 2) (filter even?)))
(defn transduce-xf-sum [coll] (transduce xf-sum + coll))

(println "Transduce with xf-sum:" (transduce-xf-sum [1 2 3 4 5]))


;; ── Memoization ─────────────────────────────────────

(println "\n--- Memoization (with-cache)")

(def memoize
  (let [cache (atom {})]
    (fn [f]
      (fn [& args]
        (if-let [result (get @cache args)]
          result
          (let [new-result (apply f args)]
            (swap! cache assoc args new-result)
            new-result)))))))


(defn slow-fib [n]
  (Thread/sleep 10))  ; Simulate slowness
  (if (< n 2)
    n
    (+ (slow-fib (dec n)) (slow-fib (dec n))))

(def fib (memoize slow-fib))

(println "fib(30) =" (fib 30))


;; ── Thread-last macro (->> pipeline) ─────────────────

(println "\n--- Thread-last macro (pipeline)")

(defn process [n]
  (* 2 n))

(def pipeline (comp process process process))

(println "pipeline(5) =" (pipeline 5))


;; ── Summary ───────────────────────────────────────

(println "\n--- Summary ---")
(println "Clojure FP features:")
(println "  - Pure functions: no side effects")
(println "  - Immutable data: vectors, maps, sets")
(println "  - Higher-order: map, filter, reduce")
(println "  - Closures: let/loop bindings")
(println "  - Partial: partial application")
(println "  - Composition: comp, juxt, partial")
(println "  - Destructuring: bind data structures")
(println "  - Lazy sequences: range, take, drop")
(println "  - Transducers: composition for reduce")
(println "  - Thread-last: ->> pipeline macro")
(println "  - Memoization: with-cache optimization")
(println "  - No loops: use map/filter/reduce instead")

File: examples/typescript/07-fp-features/main.ts
// FP Features in TypeScript
// ======================
// Arrow functions, array methods, generics, async/await

function main() {

    // ── Arrow functions ───────────────────────────────────

    console.log("--- Arrow functions ---");

    const add = (a: number, b: number): number => a + b;
    const square = (x: number): number => x * x;
    const multiply = (a: number, b: number): number => a * b;

    console.log(`add(5, 3) = ${add(5, 3)}`);
    console.log(`square(4) = ${square(4)}`);


    // ── Array methods (map, filter, reduce) ───────────────

    console.log("\n--- Array methods ---");

    const numbers: number[] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

    // map
    const squares = numbers.map(x => x * x);
    console.log("Squares:", squares);

    // filter
    const evens = numbers.filter(x => x % 2 === 0);
    console.log("Evens:", evens);

    // reduce
    const sum = numbers.reduce((acc, x) => acc + x, 0);
    console.log("Sum:", sum);

    // find
    const firstEven = numbers.find(x => x % 2 === 0);
    console.log("First even:", firstEven);

    // some/any
    const hasOdd = numbers.some(x => x % 2 !== 0);
    console.log("Has odd?", hasOdd);

    const allPositive = numbers.every(x => x > 0);
    console.log("All positive?", allPositive);


    // ── Method references ─────────────────────────────

    console.log("\n--- Method references ---");

    const str = "  hello  ";
    console.log("trim:", str.trim());
    console.log("toUpperCase:", str.toUpperCase());

    const trimToUpperCase = str.trim().toUpperCase;
    console.log("trimToUpperCase:", trimToUpperCase);


    // ── Chaining methods ───────────────────────────────

    console.log("\n--- Chaining methods ---");

    const result = numbers
        .filter(x => x % 2 === 0)
        .map(x => x * x)
        .slice(0, 3);

    console.log("First 3 even squares:", result);


    // ── Function composition ─────────────────────────────

    console.log("\n--- Function composition ---");

    const inc = (x: number) => x + 1;
    const triple = (x: number) => x * 3;

    // Manual composition
    const incThenTriple = (x: number) => triple(inc(x));
    console.log("incThenTriple(5) =", incThenTriple(5));

    // Compose helper
    function compose<A, B, C>(f: (x: A) => B, g: (y: B) => C): (x: A) => C(g(f(x)));
    const tripleThenInc = compose(triple, inc);
    console.log("compose(triple, inc)(5) =", tripleThenInc(5));


    // ── Optional chaining ───────────────────────────────

    console.log("\n--- Optional chaining ---");

    const maybeValue: number | null = Math.random() > 0.5 ? Math.floor(Math.random() * 100) : null;

    const doubled = maybeValue?.toString().repeat(2) ?? "no value";
    console.log("Doubled:", doubled);


    // ── Generics ───────────────────────────────────────

    console.log("\n--- Generics ---");

    function first<T>(items: T[]): T | undefined {
        return items[0];
    }

    function identity<T>(value: T): T {
        return value;
    }

    console.log("First of [1, 2, 3]:", first([1, 2, 3]));
    console.log("Identity(42):", identity(42));


    // ── Destructuring ───────────────────────────────

    console.log("\n--- Destructuring ---");

    const point = { x: 3, y: 4 };
    const { x, y } = point;
    console.log(`Point: (${x}, ${y})`);

    const [first, second] = [1, 2];
    console.log(`First: ${first}, Second: ${second}`);


    // ── Spread operator ─────────────────────────────

    console.log("\n--- Spread operator ---");

    function sumAll(...nums: number[]): number {
        return nums.reduce((a, b) => a + b, 0);
    }

    console.log("Sum of 1..5:", sumAll(1, 2, 3, 4, 5));

    function maxAll(...nums: number[]): number {
        return Math.max(...nums);
    }

    console.log("Max of 1, 5, 3:", maxAll(1, 5, 3));


    // ── Async/await with arrays ─────────────────────────────

    console.log("\n--- Async/await arrays ---");

    async function fetchAll(urls: string[]): Promise<string[]> {
        return Promise.all(urls.map(async url => {
            await new Promise(r => setTimeout(() => r(`fetched ${url}`), 100));
        }));
    }

    fetchAll(["url1", "url2", "url3"]).then(results => {
        console.log("Fetched:", results);
    });


    // ── Summary ─────────────────────────────────────

    console.log("\n--- Summary ---");
    console.log("TypeScript FP features:");
    console.log("  - Arrow functions: const f = (x) => {}");
    console.log("  - Array methods: map, filter, reduce, find");
    console.log("  - Method references: obj.method");
    console.log("  - Optional chaining: ?. operator");
    console.log("  - Generics: <T>(x: T)");
    console.log("  - Destructuring: const {x, y} = obj");
    console.log("  - Spread operator: ...array, ...rest");
    console.log("  - Chaining: methods().method().method()");
    console.log("  - Async/await: Promise.all, async functions");
    console.log("  - Type inference: explicit types optional");
    console.log("  - Function composition: manual or pipe libraries");
}
